src/version.rs

```diff
@@ -207,6 +215,55 @@ impl Version {
     }
 }
 
+impl NewVersion {
+    pub fn new(
+        crate_id: i32,
+        num: &semver::Version,
+        features: &HashMap<String, Vec<String>>,
+    ) -> CargoResult<Self> {
+        let features = json::encode(features)?;
+        Ok(NewVersion {
+            crate_id: crate_id,
+            num: num.to_string(),
+            features: features,
+        })
+    }
+
+    pub fn save(&self, conn: &PgConnection, authors: &[String]) -> CargoResult<Version> {
+        use diesel::{select, insert};
+        use diesel::expression::dsl::exists;
+        use schema::versions::dsl::*;
```

Comment:
do i need to check this for underhandedness? :P https://twitter.com/sgrif/status/842850645577154561